File: question/neetcode150/linked_list/reorder-list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """

        slow = head
        fast = head.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        second = slow.next
        prev = None
        while second:
            nxt = second.next
            second.next = prev
            prev = second
            second = nxt

        first, second = head, prev

        while second:
            n1, n2 = first.next, second.next
            first.next = n2
            second.next = n1
            first,second = n1,n2
        

File: question/neetcode/linked_list/reorder-list.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        void reorderList(ListNode* head) {
            ListNode* slow = head;
            ListNode* fast = head;

            while(fast and fast->next){
                slow = slow->next;
                fast = fast->next->next;
            }

            ListNode* second = slow->next;
            ListNode* prev = nullptr;
            slow->next = nullptr;

            while(second){
                ListNode* next = second->next;
                second->next = prev;
                prev = second;
                second = next;
            }

            ListNode* first = head;
            second = prev;
            while(first and second){
                ListNode* firstNext = first->next;
                ListNode* secondNext = second->next;
                first->next = second;
                second->next = firstNext;

                first = firstNext;
                second = secondNext;
            }
        }
    };